File: frontend/src/lib/discover.ts
import { requestAuthedJson } from "./auth";

export type DiscoverPlatform = "weixin" | "toutiao";
export type DiscoverTimeRange = "all" | "1d" | "3d" | "7d" | "1m" | "3m";
export type HotTopicTrend = "up" | "down" | "stable";
export type HotTopicPlatform = "weibo" | "baidu" | "toutiao" | "global";

export type Pagination = {
  page: number;
  page_size: number;
  total: number;
  has_more: boolean;
};

export type DiscoverArticle = {
  id: string;
  platform: DiscoverPlatform;
  field: string;
  title: string;
  author: string;
  publish_time: string;
  views: number;
  likes: number;
  shares: number;
  source_url?: string | null;
  is_sample: boolean;
  is_hot: boolean;
  is_new: boolean;
};

export type HotTopic = {
  id: string;
  rank: number;
  platform: HotTopicPlatform;
  title: string;
  heat: number;
  trend: HotTopicTrend;
  field: string;
  source_url?: string | null;
};

export type DiscoverArticleList = {
  items: DiscoverArticle[];
  pagination: Pagination;
  synced_at?: string | null;
};

export type DiscoverArticleRefreshResult = {
  total: number;
  synced_at?: string | null;
};

export type HotTopicList = {
  items: HotTopic[];
  pagination: Pagination;
  synced_at?: string | null;
};

export type HotTopicRefreshResult = {
  total: number;
  synced_at?: string | null;
};

export type DiscoverArticleQuery = {
  platform?: DiscoverPlatform;
  keyword?: string;
  field?: string;
  time_range?: DiscoverTimeRange;
  views_min?: number;
  page?: number;
  page_size?: number;
};

function buildQueryString(params: Record<string, string | number | undefined>) {
  const searchParams = new URLSearchParams();

  Object.entries(params).forEach(([key, value]) => {
    if (value === undefined || value === "") {
      return;
    }
    searchParams.set(key, String(value));
  });

  const queryString = searchParams.toString();
  return queryString ? `?${queryString}` : "";
}

export async function fetchDiscoverArticles(query: DiscoverArticleQuery) {
  const queryString = buildQueryString({
    platform: query.platform,
    keyword: query.keyword,
    field: query.field,
    time_range: query.time_range,
    views_min: query.views_min,
    page: query.page,
    page_size: query.page_size,
  });
  return requestAuthedJson<DiscoverArticleList>(`/api/v1/discover/articles${queryString}`);
}

export async function refreshDiscoverArticlesSnapshot() {
  return requestAuthedJson<DiscoverArticleRefreshResult>("/api/v1/discover/articles", {
    method: "POST",
  });
}

export async function fetchHotTopics(params: { page?: number; page_size?: number }) {
  const queryString = buildQueryString({
    page: params.page,
    page_size: params.page_size,
  });
  return requestAuthedJson<HotTopicList>(`/api/v1/discover/hot-topics${queryString}`);
}

export async function refreshHotTopicsSnapshot() {
  return requestAuthedJson<HotTopicRefreshResult>("/api/v1/discover/hot-topics", {
    method: "POST",
  });
}

File: backend/app/schemas/discover.py
from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict


class DiscoverTimeRange(str, Enum):
    ALL = "all"
    ONE_DAY = "1d"
    THREE_DAYS = "3d"
    SEVEN_DAYS = "7d"
    ONE_MONTH = "1m"
    THREE_MONTHS = "3m"


class Pagination(BaseModel):
    page: int
    page_size: int
    total: int
    has_more: bool


class DiscoverArticleItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    platform: str
    field: str
    title: str
    author: str
    publish_time: datetime
    views: int
    likes: int
    shares: int
    source_url: str | None = None
    is_sample: bool = False
    is_hot: bool
    is_new: bool


class DiscoverArticleList(BaseModel):
    items: list[DiscoverArticleItem]
    pagination: Pagination


class HotTopicItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    rank: int
    platform: str
    title: str
    heat: int
    trend: str
    field: str
    source_url: str | None = None


class HotTopicList(BaseModel):
    items: list[HotTopicItem]
    pagination: Pagination
    synced_at: datetime | None = None


class HotTopicRefreshResult(BaseModel):
    total: int
    synced_at: datetime | None = None


class DiscoverArticleQuery(BaseModel):
    platform: str | None = None
    keyword: str | None = None
    field: str | None = None
    time_range: DiscoverTimeRange = DiscoverTimeRange.ALL
    views_min: int | None = None
    page: int = 1
    page_size: int = 20
